### DISCUSSION TITLE
Can we allow pointers to substructures in 7.1?

### DISCUSSION THREAD

**Discussion by tychonievich** [FamilySearch/GEDCOM#328] [Ideas]
7.0 prohibits pointers to substructures as follows:

> The cross-reference identifier [...] indicates that this is a structure to which pointer-type payloads may point. 
> [...]
> A substructure [...] must not have a cross-reference identifier.

I'd like to see the prohibition on pointers to substructures removed from 7.1. A motivating example follows.

---

I have an extension that uses pointers to substructures to track the dependencies of reasoning. I use this to document complicated reasoning in cases where names are not enough to identify individuals, with a machine-readable tracking of the dependencies of my reasoning.

For example, I might pick which of two people is the bride in a marrriage based on their age. Given

```gedcom
0 @I1@ INDI
1 NAME Jane /Potter/
1 BIRT
2 DATE 3 JUN 1845
0 @I2@ INDI
1 NAME Jane /Potter/
1 BIRT
2 DATE 21 SEP 1856
0 @F1@ FAM
1 MARR
2 DATE 19 JAN 1866
2 PLAC Albany, Albany, New York, United States of America
2 SOUR @S1@
3 DATA
4 TEXT ...
5 CONT Groom: Elias Faux
5 CONT Bride: Jane Potter
5 CONT ...
```

I'd add a `WIFE` structure with supporting reasoning, like so

```gedcom
0 @I1@ INDI
1 @N1@ NAME Jane /Potter/
1 BIRT
2 @D1@ DATE 3 JUN 1845
0 @I2@ INDI
1 @N2@ NAME Jane /Potter/
1 BIRT
2 @D2@ DATE 21 SEP 1856
0 @F1@ FAM
1 MARR
1 WIFE @I1@
2 PHRASE Jane Potter
2 _REASON @R1@
2 @D3@ DATE 19 JAN 1867
2 @P1@ PLAC Albany, Albany, New York, United States of America
2 SOUR @S1@
3 DATA
4 TEXT ...
5 CONT Groom: Elias Faux
5 CONT Bride: Jane Potter
5 CONT ...
0 @R1@ _REASONING
1 _SUPPORT @N1@
1 _SUPPORT @N2@
1 _SUPPORT @D1@
1 _SUPPORT @D2@
1 _SUPPORT @D3@
1 _SUPPORT @P1@
1 _ASSUMPTION in 19th century New York, 10-year-olds are not married
1 _ASSUMPTION there are no other Jane Potters to consider for this marriage
1 NOTE I found two candidate Jane Potters who match all my other data.
2 CONT Of the two, I'm picking the 21y old over the 10y old.
```

However, this violates the rule that substructures can't have cross-reference identifiers, and thus cannot be pointed to by other structures.

I can work around this by making a record copy of everything I might want to reference (all events, names, dates, places, etc)
but that results in a large amount of duplicated data and a much more complicated set of extension structure types. I could also just remove the `_SUPPORT` structure altogether, but then I can't have my tool automatically warn me when a change I make invalidates some of my reasoning, nor can I communicate the essence of my reasoning in a machine-readable way.

If 7.1 adds support for pointers to substructures, I'll be able to use this extension without violating the standard or creating a few hundred extension types and significant data duplication.

**Comment by dthaler** [comment]
I think you can do this (differently) in 7.0 today using extensions, e.g. something like:
```
0 @I1@ INDI
1 NAME Jane /Potter/
1 _PTR N1
1 BIRT
2 DATE 3 JUN 1845
3 _PTR D1
0 @I2@ INDI
1 NAME Jane /Potter/
1 _PTR N2
1 BIRT
2 DATE 21 SEP 1856
3 _PTR D2
0 @F1@ FAM
1 MARR
1 WIFE @I1@
2 PHRASE Jane Potter
2 _REASON @R1@
2 DATE 19 JAN 1867
3 _PTR D3
2 PLAC Albany, Albany, New York, United States of America
3 _PTR P1
2 SOUR @S1@
3 DATA
4 TEXT ...
5 CONT Groom: Elias Faux
5 CONT Bride: Jane Potter
5 CONT ...
0 @R1@ _REASONING
1 _SUPPORT N1
1 _SUPPORT N2
1 _SUPPORT D1
1 _SUPPORT D2
1 _SUPPORT D3
1 _SUPPORT P1
1 _ASSUMPTION in 19th century New York, 10-year-olds are not married
1 _ASSUMPTION there are no other Jane Potters to consider for this marriage
1 NOTE I found two candidate Jane Potters who match all my other data.
2 CONT Of the two, I'm picking the 21y old over the 10y old.
```

**Reply by tychonievich** [reply]
An interesting proposal. Change all

```
2 @ID@ TAG
1 _SUPPORT @ID@
```

to 

```
2 TAG
3 _PTR ID
1 _SUPPORT ID
```

That does avoid record proliferation, but it also makes for datasets that cannot be merged without understanding the semantics of the extension, which I am very hesitant to do. If I had to do this I'd rather do it with something that can be merged, like UID, but even that feels like a hack: we have pointers in the spec, with a syntax that allows them anywhere and just one half sentence barring that. I'd prefer to remove that half of that sentence and do this (and other similar ideas, such as the ALIA-like pointers inside events that some applications have) using the existing GEDCOM pointer mechanism.

**Comment by albertemmerich** [comment]
I have a different approach:
An INDI record represents a real existing individual. This individual cannot have two different birth dates. So I have two INDI @I1@ and @I2@ for Jane /Potter/, one for the Jane born in 1845, one for the Jane born in 1856. These are obviously 2 different persons. I have a third event: the marriage. As long as I am not sure who is the bride I have a third INDI @I3@ used with the WIFE tag in the FAM record.
I can do reasoning using these pointers @I1@, @I2@, @I3@. And I can show the possible coincidence using the ALIA tag.
Merging @I3@ to @I1@ (or @I2@) will be done if I have a conclusion, which one is the bride. But I never will put the two birth events into one INDI record - two persons need two records!

Saying this I do not see the example telling us we need pointers to substructures.

**Reply by tychonievich** [reply]
This approach (or the version using ALIA instead of merging) is a nice way of handling the uncertainty, but does not capture the dependency between a conclusion (such as the merge action) and the supporting evidence (such as the name and birth date of the persons merged). If a subsequent merge with another INDI @i4@ overrode the name or birth date, substructure pointers allow automatic notification that previous reasoning may no longer apply.

**Comment by tychonievich** [comment]
A few other example reasons I want pointers to substructures:

- A source lists a person under their professional name instead of their legal name. I want a `_USING_NAME @NAME2@` in the source pointing to the `NAME`.`TYPE PROFESSIONAL`.
- An extension substructure overlaps in meaning with a standard substructure enough that the two are supposed to be kept in sync. I want a `_SYNC_WITH @STD1@` as a substructure of the extension.
- I have a source stating that two people lived in a particular place at the same time, but when that was. I want a `_CONCURRENTLY @OTHER_RESI@` in each INDI's RESI.
- A child was named after the place where the parents resided when conceived. I want a `_NAMED_FOR @PLACE1@` pointing to the `FAM`.`RESI`.`PLAC`. This can partly be done with extension location records, but the because-they-resided-there can't.
- Two events look similar, but I have evidence they are distinct. I want a `_NOT_SAME_AS @EVENT2@` substructure inside event 1. This can also be done with extension event records, at the cost of data duplication.

.. and so on. I find myself encountering this desire repeatedly as I work on family history applications.

Admittedly, all of these can be solved in other ways. I could make record copies of the pointed-to thing and point the substructures to the records. I could make a _XREF_ID substructure to encode pointers without using pointers. I could use UUIDs to make the  _XREF_ID substructure resistant to collisions during file merges. If the decision is "No, 7.x will continue to ban pointers to substructures even in extensions" I can survive. But all of the work-arounds I've found so far involve duplicating information or creating two alternate ways to expressing the same information, which feel like anti-patterns from a software engineering perspective; thus I'd rather just add pointers to substructures.

**Comment by Norwegian-Sardines** [comment]
This topic exposes the bigger issue in that GEDCOM is not normalized very well.  If a place can be used in multiple facts and/or substructures it needs to be broken out as a new (and referenced) entity.  

The same is true for source_citations, events, surnames and probably other structures that are regularly reused.  

This also brings to mind the talk/suggestions for changes to the level 1 FAMC linkage to include pedigree information about the parents, removing them from the BIRT and ADOP events.  Normalization rules would indicate that these need to be linkable records so multiple structures and use the data, rather than having the data in multiple places. 

This suggests that some of these changes are better held off for GEDCOM v8.0, so that other normalization concepts (a FACT/EVEN Record) can be reviewed.  The concept of inner record linkage looks great to reduce redundancy, but from a database design perspective (SQL, Relational) it is not.  If we were talking about object oriented databases exclusively then it has more benefits.

**Reply by tychonievich** [reply]
These are good points. I'm now leaning toward waiting for 8.0 and rethink how this works at a broader level.
